### PR TITLE
Add complex64 and complex128 support for ApplyAdadelta kernel

### LINKED ISSUE #13521: complex gradient update in optimization
Is there a plan to allow complex optimization in Tensorflow in the future?

When you try to do it with version 1.3, you can calculate and evaluate gradients, but you cannot apply them with opt.apply_gradients(grds_and_vars). The error message you get is: 

InvalidArgumentError (see above for traceback): No OpKernel was registered to support Op 'ApplyAdadelta' with these attrs.  Registered devices: [CPU,GPU], Registered kernels:
  device='GPU'; T in [DT_DOUBLE]
  device='GPU'; T in [DT_FLOAT]
  device='GPU'; T in [DT_HALF]
  device='CPU'; T in [DT_DOUBLE]
  device='CPU'; T in [DT_FLOAT]
  device='CPU'; T in [DT_HALF]

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
No

- **OS Platform and Distribution 
Distributor ID: Debian
Description:    Debian GNU/Linux 8.9 (jessie)
Release:        8.9

- **TensorFlow installed from (source or binary)**:
using pip install in a virtual conda environment

- **TensorFlow version (use command below)**:
1.3

- **Python version**: 
Python 3.5.2 |Anaconda 4.3.0 (64-bit)| (default, Jul  2 2016, 17:53:06) 

- **Bazel version (if compiling from source)**:
not applicable

- **CUDA/cuDNN version**:
8.0/6.0

- **GPU model and memory**:
Tesla K40c, 11439MiB

- **Exact command to reproduce**:
not necessary, since feature request/question

### Describe the problem
problem description above

### Source code / logs
not applicable

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #13521 where the complex64 and complex128 support for ApplyAdadelta is missing. The test cases have also been updated.

This fix fixes #13521.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.cc [hunk 1]
```diff
@@ -649,6 +649,8 @@ class ApplyAdadeltaOp : public OpKernel {
 TF_CALL_half(REGISTER_CPU_KERNELS);
 TF_CALL_float(REGISTER_CPU_KERNELS);
 TF_CALL_double(REGISTER_CPU_KERNELS);
+TF_CALL_complex64(REGISTER_CPU_KERNELS);
+TF_CALL_complex128(REGISTER_CPU_KERNELS);
 
 #if GOOGLE_CUDA
 // Forward declarations of the functor specializations for GPU.
```
Comment (ebrevdo) [inline tensorflow/core/kernels/training_ops.cc]: No GPU support?

**Comment by yongtang** [comment]
@ebrevdo Thanks for the review. I tried to enable complex for GPU but have the following `host function("__builtin_csqrtf") from a device function` error.

I am guessing Eigen does not have a complex version of `sqrt` on GPU yet.

Maybe we could leave GPU part alone for now?

If I could be pointed to the place in Eigen to make the necessary change, I will be happy to submit a PR in Eigen.

external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorEvaluator.h(420): error: calling a __host__ function("__builtin_csqrtf") from a __device__ function("Eigen::internal::EigenMetaKernelEval< ::Eigen::TensorEvaluator<const  ::Eigen::TensorAssignOp< ::Eigen::TensorMap< ::Eigen::Tensor< ::std::complex<float> , (int)1, (int)1, long> , (int)16,  ::Eigen::MakePointer> , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_difference_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorMap< ::Eigen::Tensor< ::std::complex<float> , (int)1, (int)1, long> , (int)16,  ::Eigen::MakePointer> , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_product_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_product_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_product_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorCwiseUnaryOp< ::Eigen::internal::scalar_sqrt_op< ::std::complex<float> > , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_sum_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorMap< ::Eigen::Tensor< ::std::complex<float> , (int)1, (int)1, long> , (int)16,  ::Eigen::MakePointer> , const  ::Eigen::TensorBroadcastingOp<const  ::Eigen::array<long, (unsigned long)1ul> , const  ::Eigen::TensorReshapingOp<const  ::Eigen::Sizes<(long)1l > , const  ::Eigen::TensorMap< ::Eigen::TensorFixedSize<const  ::std::complex<float> ,  ::Eigen::Sizes< > , (int)1, long> , (int)16,  ::Eigen::MakePointer> > > > > , const  ::Eigen::TensorCwiseUnaryOp< ::Eigen::internal::scalar_rsqrt_op< ::std::complex<float> > , const  ::Eigen::TensorCwiseBinaryOp< ::Eigen::internal::scalar_sum_op< ::std::complex<float> ,  ::std::complex<float> > , const  ::Eigen::TensorMap< ::Eigen::Tensor< ::std::complex<float> , (int)1, (int)1, long> , (int)16,  ::Eigen::MakePointer> , const  ::Eigen::TensorBroadcastingOp<const  ::Eigen::array<long, (unsigned long)1ul> , const  ::Eigen::TensorReshapingOp<const  ::Eigen::Sizes<(long)1l > , const  ::Eigen::TensorMap< ::Eigen::TensorFixedSize<const  ::std::complex<float> ,  ::Eigen::Sizes< > , (int)1, long> , (int)16,  ::Eigen::MakePointer> > > > > > , const  ::Eigen::TensorMap< ::Eigen::Tensor<const  ::std::complex<float> , (int)1, (int)1, long> , (int)16,  ::Eigen::MakePointer> > , const  ::Eigen::TensorBroadcastingOp<const  ::Eigen::array<long, (unsigned long)1ul> , const  ::Eigen::TensorReshapingOp<const  ::Eigen::Sizes<(long)1l > , const  ::Eigen::TensorMap< ::Eigen::TensorFixedSize<const  ::std::complex<float> ,  ::Eigen::Sizes< > , (int)1, long> , (int)16,  ::Eigen::MakePointer> > > > > > ,  ::Eigen::GpuDevice> , long, (bool)0> ::run") is not allowed

**Comment by rmlarsen** [comment]
I'm guessing that you are missing an .abs() or .abs2() somewhere. There is almost guaranteed no use for complex square root.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by rmlarsen** [comment]
@yongtang any luck debugging this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
Sorry for the delay. Missed the notification. I will work on this issue in the next few days and hopefully get it addressed shortly.

**Comment by yongtang** [comment]
Let me close this PR for now. I will reopen once I figure out the way to fix it.

Sorry to all for the inconvenience.

**Pull request closed without merging**